unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WalletHUD/WalletCardHUDController.cs

```diff
@@ -0,0 +1,127 @@
+        private readonly ITheGraph theGraph;
+        private readonly DataStore dataStore;
+
+        private CancellationTokenSource fetchEthereumManaCancellationToken;
+        private CancellationTokenSource fetchPolygonManaCancellationToken;
+
+        private UserProfile ownUserProfile => userProfileBridge.GetOwn();
+
+        public WalletCardHUDController(
+            IWalletCardHUDComponentView view,
+            IUserProfileBridge userProfileBridge,
+            ITheGraph theGraph,
+            DataStore dataStore)
+        {
+            this.view = view;
+            this.userProfileBridge = userProfileBridge;
+            this.theGraph = theGraph;
+            this.dataStore = dataStore;
+
+            dataStore.wallet.currentEthereumManaBalance.OnChange += OnCurrentEthereumManaBalanceChanged;
+            OnCurrentEthereumManaBalanceChanged(dataStore.wallet.currentEthereumManaBalance.Get(), 0f);
+            dataStore.wallet.currentPolygonManaBalance.OnChange += OnCurrentPolygonManaBalanceChanged;
+            OnCurrentPolygonManaBalanceChanged(dataStore.wallet.currentPolygonManaBalance.Get(), 0f);
+
+            dataStore.wallet.isWalletCardVisible.OnChange += OnWalletCardVisible;
+        }
+
+        public void Dispose()
+        {
+            dataStore.wallet.currentEthereumManaBalance.OnChange -= OnCurrentEthereumManaBalanceChanged;
+            dataStore.wallet.currentPolygonManaBalance.OnChange -= OnCurrentPolygonManaBalanceChanged;
+            dataStore.wallet.isWalletCardVisible.OnChange -= OnWalletCardVisible;
+
+            fetchEthereumManaCancellationToken.SafeCancelAndDispose();
+            fetchPolygonManaCancellationToken.SafeCancelAndDispose();
+        }
+
+        private void OnCurrentEthereumManaBalanceChanged(double currentBalance, double _) =>
+            view.SetEthereumManaBalance(currentBalance);
+
+        private void OnCurrentPolygonManaBalanceChanged(double currentBalance, double _) =>
+            view.SetPolygonManaBalance(currentBalance);
+
+        private void OnWalletCardVisible(bool isVisible, bool _)
+        {
+            if (ownUserProfile.isGuest)
+                return;
+
+            if (isVisible)
+            {
+                fetchEthereumManaCancellationToken = fetchEthereumManaCancellationToken.SafeRestart();
+                fetchPolygonManaCancellationToken = fetchPolygonManaCancellationToken.SafeRestart();
+                RequestEthereumManaAsync(fetchEthereumManaCancellationToken.Token).Forget();
+                RequestPolygonManaAsync(fetchPolygonManaCancellationToken.Token).Forget();
+            }
+            else
+            {
+                fetchEthereumManaCancellationToken.SafeCancelAndDispose();
+                fetchPolygonManaCancellationToken.SafeCancelAndDispose();
+            }
+        }
+
+        private async UniTask RequestEthereumManaAsync(CancellationToken cancellationToken)
+        {
+            while (true)
```

Comment:
I am not sure about this infinite loop here. Can it lead to the situation when will infinitely hang? Maybe we can limit it by retry counts inside this loop condition?